File: surf-api-bukkit/surf-api-bukkit-server/src/main/java/dev/slne/surf/surfapi/bukkit/server/scoreboard/SurfScoreboardBuilderImpl.java
package dev.slne.surf.surfapi.bukkit.server.scoreboard;

import com.google.common.collect.Lists;
import dev.slne.surf.surfapi.bukkit.api.scoreboard.SurfAutoUpdatablePlayerScoreboard;
import dev.slne.surf.surfapi.bukkit.api.scoreboard.SurfAutoUpdatableScoreboard;
import dev.slne.surf.surfapi.bukkit.api.scoreboard.SurfScoreboard;
import dev.slne.surf.surfapi.bukkit.api.scoreboard.SurfScoreboardBuilder;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.TextColor;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import net.megavex.scoreboardlibrary.api.sidebar.component.SidebarComponent;
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.CollectionSidebarAnimation;
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.SidebarAnimation;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

import static com.google.common.base.Preconditions.*;

@ApiStatus.Internal
public class SurfScoreboardBuilderImpl implements SurfScoreboardBuilder {

    private final Component title;
    private final SidebarComponent.Builder sidebarComponentBuilder;
    private final List<CollectionSidebarAnimation<Component>> animations = new ArrayList<>();
    private int maxLines = DEFAULT_MAX_LINES;

    public SurfScoreboardBuilderImpl(@NotNull Component title) {
        checkNotNull(title, "title");

        this.title = title;
        this.sidebarComponentBuilder = SidebarComponent.builder();
    }

    @Override
    public SurfScoreboardBuilder maxLines(int maxLines) {
        checkArgument(maxLines >= 1 && maxLines <= 15, "maxLines must be between 1 and 15");

        this.maxLines = maxLines;
        return this;
    }

    @Override
    public SurfScoreboardBuilder addLine(@NotNull Component line) {
        checkNotNull(line, "line");

        sidebarComponentBuilder.addStaticLine(line);
        return this;
    }

    @Override
    public SurfScoreboardBuilder addUpdatableLine(@NotNull Supplier<@NotNull Component> line) {
        checkNotNull(line, "line");

        sidebarComponentBuilder.addDynamicLine(line);
        return this;
    }

    @Override
    public SurfScoreboardBuilder addAnimatedLine(@NotNull SidebarAnimation<@NotNull SidebarComponent> animation) {
        checkNotNull(animation, "animation");
        
        sidebarComponentBuilder.addAnimatedComponent(animation);
        return this;
    }

    @Override
    public SurfScoreboardBuilder addAnimatedLine(@NotNull List<@NotNull Component> frames) {
        checkNotNull(frames, "frames");
        checkArgument(!frames.isEmpty(), "frames cannot be empty");

        CollectionSidebarAnimation<@NotNull Component> animation = new CollectionSidebarAnimation<>(frames);
        sidebarComponentBuilder.addAnimatedLine(animation);
        animations.add(animation);
        return this;
    }

    @Override
    public SurfScoreboardBuilder addGradientLine(@NotNull Component text, @NotNull TextColor start, @NotNull TextColor end) {
        checkNotNull(text, "text");
        checkNotNull(start, "start");
        checkNotNull(end, "end");

        sidebarComponentBuilder.addAnimatedLine(createGradientAnimation(text, start.asHexString(), end.asHexString()));
        return this;
    }

    @Override
    public @NotNull SurfScoreboard build() {
        return new SurfScoreboardImpl(title, maxLines, sidebarComponentBuilder.build(), animations);
    }

    @Override
    public @NotNull SurfAutoUpdatableScoreboard buildAutoUpdatable() {
        return new SurfAutoUpdatableScoreboardImpl(title, maxLines, sidebarComponentBuilder.build(), animations);
    }

    @Override
    public @NotNull SurfAutoUpdatablePlayerScoreboard buildAutoUpdatablePlayer() {
        return new SurfAutoUpdatablePlayerScoreboardImpl(title, maxLines, sidebarComponentBuilder.build(), animations);
    }

    @Contract("_, _, _ -> new")
    private static @NotNull SidebarAnimation<Component> createGradientAnimation(Component component, String firstHex, String secondHex) {
        final float step = 1f / 20f;
        final TagResolver.Single textPlaceholder = Placeholder.component("text", component);
        final List<Component> frames = new ArrayList<>();

        // Animation from left to right
        var phase = -1f;
        while (phase < 1) {
            frames.add(
                    MiniMessage.miniMessage().deserialize("<gradient:%s:%s:%s><text></gradient>".formatted(firstHex, secondHex, phase), textPlaceholder)
            );
            phase += step;
        }

        // Animation from right to left
        frames.addAll(Lists.reverse(frames));

        return new CollectionSidebarAnimation<>(frames);
    }
}

File: surf-api-paper/surf-api-paper-server/src/main/kotlin/dev/slne/surf/api/paper/server/impl/scoreboard/SurfScoreboardBuilderImpl.kt
package dev.slne.surf.api.paper.server.impl.scoreboard

import dev.slne.surf.api.core.util.mutableObjectListOf
import dev.slne.surf.api.paper.scoreboard.SurfScoreboardBuilder
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextColor
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import net.megavex.scoreboardlibrary.api.sidebar.component.SidebarComponent
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.CollectionSidebarAnimation
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.SidebarAnimation
import org.bukkit.entity.Player
import java.util.function.Function
import java.util.function.Supplier

class SurfScoreboardBuilderImpl(private val title: Component) : SurfScoreboardBuilder {
    private val lines = mutableObjectListOf<ScoreboardLine>()
    private val snapshots = mutableObjectListOf<LineSnapshot<*>>()
    private val animations = mutableObjectListOf<SidebarAnimation<Component>>()
    private var maxLines = SurfScoreboardBuilder.DEFAULT_MAX_LINES

    override fun maxLines(maxLines: Int) = apply {
        require(maxLines in 1..15) { "maxLines must be between 1 and 15" }
        this.maxLines = maxLines
    }

    override fun addLine(line: Component) = addShared(SidebarComponent.staticLine(line))

    override fun addUpdatableLine(line: Supplier<Component>) = apply {
        val snapshot = LineSnapshot(line::get).also { snapshots.add(it) }
        addShared { drawable -> drawable.drawLine(snapshot.value) }
    }

    override fun addViewerComponent(component: Function<Player, SidebarComponent>) = apply {
        lines.add(ScoreboardLine.Viewer(component))
    }

    override fun addAnimatedLine(animation: SidebarAnimation<SidebarComponent>) =
        addShared(SidebarComponent.animatedComponent(animation))

    override fun addAnimatedLine(frames: MutableList<Component>) = apply {
        check(frames.isNotEmpty()) { "frames cannot be empty" }
        addAnimation(CollectionSidebarAnimation(frames))
    }

    override fun addGradientLine(text: Component, start: TextColor, end: TextColor) =
        addAnimation(createGradientAnimation(text, start.asHexString(), end.asHexString()))

    private fun addAnimation(animation: SidebarAnimation<Component>) = apply {
        animations.add(animation)
        addShared(SidebarComponent.animatedLine(animation))
    }

    private fun addShared(component: SidebarComponent) = apply {
        lines.add(ScoreboardLine.Shared(component))
    }

    private fun definition() = ScoreboardDefinition(
        title,
        maxLines,
        lines.toList(),
        snapshots.toList(),
        animations.toList(),
    )

    override fun build() = SurfScoreboardImpl(definition())
    override fun buildAutoUpdatable() = SurfAutoUpdatableScoreboardImpl(definition())
    override fun buildAutoUpdatablePlayer() = SurfAutoUpdatablePlayerScoreboardImpl(definition())

    companion object {
        private fun createGradientAnimation(
            component: Component,
            firstHex: String, secondHex: String
        ): SidebarAnimation<Component> {
            val step = 1f / 20f
            val textPlaceholder = Placeholder.component("text", component)
            val frames = mutableObjectListOf<Component>()

            // Animation from left to right
            var phase = -1f
            while (phase < 1) {
                frames.add(
                    MiniMessage.miniMessage().deserialize(
                        "<gradient:$firstHex:$secondHex:$phase><text></gradient>",
                        textPlaceholder
                    )
                )
                phase += step
            }

            // Animation from right to left
            frames.addAll(frames.reversed())

            return CollectionSidebarAnimation(frames)
        }
    }
}
